File: frontend/src/lib/eventCode.js
/**
 * Short, human-copyable event codes.
 *
 * The mirror of `com.crowdease.yasss.model.EventCode`. Two implementations of
 * the same rule is a drift hazard, so a shared table of cases is asserted
 * against both — see `tests/unit/eventCode.test.js` and `EventCodeTest.java`,
 * the same arrangement `CredentialInteropTest` uses to pin crypto across
 * languages.
 *
 * Crockford Base32, whose alphabet omits I, L, O and U and whose decoding folds
 * the characters people confuse into the ones that survive. Reading a code
 * aloud and writing it down cannot produce a different code.
 */

/** The encoding alphabet. Note the absent I, L, O and U. */
const ALPHABET = '0123456789ABCDEFGHJKMNPQRSTVWXYZ';

/** How many symbols a code has. */
export const CODE_LENGTH = 8;

/**
 * Puts a user-supplied spelling of a code into canonical form.
 *
 * Uppercase, fold the ambiguous characters, then drop everything outside the
 * alphabet — which is what makes separators and stray punctuation irrelevant.
 * `abcd-efgh`, `ABCD EFGH` and `a.b.c.d.e.f.g.h` are all one code.
 *
 * @param {unknown} raw the code as typed, pasted or read out
 * @returns {string|null} the canonical form, or null if it is not a code
 */
export function normalizeCode(raw) {
  if (typeof raw !== 'string') return null;

  let out = '';
  for (const char of raw.toUpperCase()) {
    const folded = char === 'O' ? '0' : (char === 'I' || char === 'L' ? '1' : char);
    if (ALPHABET.includes(folded)) out += folded;
  }
  return out.length === CODE_LENGTH ? out : null;
}

/**
 * Whether a string is a code in any spelling.
 *
 * @param {unknown} raw
 * @returns {boolean}
 */
export const isCode = (raw) => normalizeCode(raw) !== null;

/**
 * Renders a canonical code for display.
 *
 * The hyphen is presentation only — it is never stored and never queried.
 *
 * @param {unknown} code
 * @returns {string|null} the code as `XXXX-XXXX`, or null if there is none
 */
export function formatCode(code) {
  const canonical = normalizeCode(code);
  return canonical === null ? null : `${canonical.slice(0, 4)}-${canonical.slice(4)}`;
}

File: src/main/java/com/crowdease/yasss/model/EventCode.java
/*
 * Copyright (c) 2024 CrowdEase, LLC.
 *
 * This Source Code Form is subject to the terms of the Mozilla Public License,
 * v. 2.0. If a copy of the MPL was not distributed with this file, You can
 * obtain one at https://mozilla.org/MPL/2.0/.
 */
package com.crowdease.yasss.model;

import java.security.SecureRandom;

/**
 * Short, human-copyable identifiers for events.
 *
 * <p>Eight characters, shown as {@code XXXX-XXXX}, usable anywhere an event's
 * {@link java.util.UUID} is. The point is the copying: a UUID is thirty-six
 * characters of hex and nobody reads one down a telephone or writes one on a
 * whiteboard.
 *
 * <h2>Alphabet</h2>
 *
 * <p>Crockford Base32 — {@code 0123456789ABCDEFGHJKMNPQRSTVWXYZ} — rather than
 * anything invented here. It already specifies exactly what is wanted: the
 * encoding alphabet omits {@code I}, {@code L}, {@code O} and {@code U}, and
 * decoding folds the characters people confuse into the ones that survive, so
 * {@code O} reads as zero and both {@code I} and {@code L} read as one. Reading
 * a code aloud and writing it down cannot produce a different code.
 *
 * <p>Excluding {@code U} is Crockford's, and it is worth keeping on a public
 * service: it makes an accidental obscenity in a randomly generated code very
 * much less likely.
 *
 * <p>Eight symbols is forty bits, about 1.1 trillion codes.
 *
 * <h2>Canonical form</h2>
 *
 * <p>Uppercase, no separator. That is what is stored and what is queried, so
 * the database's collation never enters into whether two codes are the same.
 * The hyphen exists only for display.
 *
 * @author Caleb L. Power &lt;cpower@crowdease.com&gt;
 */
public final class EventCode {

  /** The encoding alphabet. Note the absent I, L, O and U. */
  private static final char[] ALPHABET = "0123456789ABCDEFGHJKMNPQRSTVWXYZ".toCharArray();

  /** How many symbols a code has. */
  public static final int LENGTH = 8;

  private static final SecureRandom RANDOM = new SecureRandom();

  private EventCode() { }

  /**
   * Puts a user-supplied spelling of a code into canonical form.
   *
   * <p>In order: uppercase; fold the ambiguous characters; drop everything that
   * is not in the alphabet, which is what makes separators and stray
   * punctuation irrelevant. {@code abcd-efgh}, {@code ABCD EFGH},
   * {@code a.b.c.d.e.f.g.h} and {@code abcdefgh} are all the same code.
   *
   * <p>A character that is neither in the alphabet nor an ambiguity — {@code U}
   * is the only one — is dropped like any other stray, which leaves the result
   * too short and so invalid. That is the right outcome: it fails rather than
   * silently resolving to a different event.
   *
   * @param raw the code as the user typed, pasted or read it out
   * @return the canonical form, or {@code null} if it is not a code
   */
  public static String normalize(String raw) {
    if(null == raw) return null;

    StringBuilder sb = new StringBuilder(LENGTH);
    for(char c : raw.toUpperCase().toCharArray()) {
      char folded = switch(c) {
        case 'O' -> '0';
        case 'I', 'L' -> '1';
        default -> c;
      };
      if(0 <= indexOf(folded)) sb.append(folded);
    }

    return LENGTH == sb.length() ? sb.toString() : null;
  }

  /**
   * Whether a string is a code in any spelling.
   *
   * @param raw the candidate
   * @return {@code true} if {@link #normalize(String)} would accept it
   */
  public static boolean isCode(String raw) {
    return null != normalize(raw);
  }

  /**
   * Renders a canonical code for display.
   *
   * @param code the canonical code
   * @return the code as {@code XXXX-XXXX}, or {@code null} if there is none
   */
  public static String format(String code) {
    String canonical = normalize(code);
    if(null == canonical) return null;
    return canonical.substring(0, 4) + '-' + canonical.substring(4);
  }

  /**
   * Generates a fresh code.
   *
   * <p>{@link SecureRandom} rather than {@code Math.random}: a code is a
   * capability in the sense that anyone holding one can read the event, so
   * guessable codes would let someone walk the space. Forty bits is not a
   * secret, but it should not be a sequence either.
   *
   * @return a new canonical code
   */
  public static String generate() {
    StringBuilder sb = new StringBuilder(LENGTH);
    for(int i = 0; i < LENGTH; i++)
      sb.append(ALPHABET[RANDOM.nextInt(ALPHABET.length)]);
    return sb.toString();
  }

  private static int indexOf(char c) {
    for(int i = 0; i < ALPHABET.length; i++)
      if(ALPHABET[i] == c) return i;
    return -1;
  }
}
